app/models/push_notification_target.rb

```diff
@@ -0,0 +1,113 @@
+class PushNotificationTarget < ApplicationRecord
+  resourcify
+
+  FOREM_BUNDLE = "com.forem.app".freeze
+  SUPPORTED_PLATFORMS = [Device::ANDROID, Device::IOS].freeze
+  FOREM_APP_PLATFORMS = [Device::IOS].freeze
+
+  validates :app_bundle, presence: true
+  validates :platform, presence: true
+
+  # Clear Redis-backed model to ensure it will be recreated with updated values
+  after_update :clear_rpush_app
+
+  def forem_app?
+    app_bundle == FOREM_BUNDLE
+  end
+
+  def active?
+    # TRUE if it's marked as `active` in the DB && it has credentials available
+    active && auth_credentials?
+  end
+
+  def auth_credentials?
+    auth_credentials.present?
+  end
+
+  def auth_credentials
+    if app_bundle == FOREM_BUNDLE && platform == Device::IOS
+      ApplicationConfig["RPUSH_IOS_PEM"]
+    else
+      auth_key
+    end
+  end
+
+  def recreate_ios_app!
+    app = Rpush::Apns2::App.new
+    app.name = app_bundle
+    app.certificate = auth_credentials.to_s.gsub("\\n", "\n")
+    app.environment = Rails.env.production? ? "production" : "development"
+    app.password = ""
+    app.bundle_id = app_bundle
+    app.connections = 1
+    app.save!
+    app
+  end
+
+  def recreate_android_app!
+    app = Rpush::Gcm::App.new
+    app.name = app_bundle
+    app.auth_key = auth_credentials.to_s
+    app.connections = 1
+    app.save!
+    app
+  end
+
+  def self.fetch_by(app_bundle:, platform:)
+    # This guard clause handles the Forem app special case
+    return forem_app_target(platform: platform) if app_bundle == FOREM_BUNDLE
+
+    # All other PushNotificationTarget are simply fetched as usual
+    PushNotificationTarget.find_by(app_bundle: app_bundle, platform: platform)
+  end
```

Comment:
I'm a bit wary of adding custom fetching methods to the model. What about a dedicated service?

1. `fetch_by`
2. `all_targets`
3. `forem_app_target`

all to me look like they should belong in service business logic.

We are trying to "hijack" a bit the logic mixing a permission problem and the model's responsibilities.

What if we combine permission policy and dedicated services for fetching, creating missing targets and loading the "forem app target"

Reply:
It's still tricky for me to identify when some logic is worth extracting into a service, but it makes total sense based on what you mention here.